### PR TITLE
Use vs-mef for OOP compositions

### PR DESCRIPTION
**Pull request by sharwell**
Microsoft.VisualStudio.Composition (vs-mef) has improved thread safety and performance relative to MEF 1 and MEF 2, plus it's the container we use for all of our testing. This change updates production scenarios to match our expectations from test scenarios.

### CODE REVIEW

File: src/Workspaces/Remote/Core/Microsoft.CodeAnalysis.Remote.Workspaces.csproj [hunk 1]
```diff
@@ -47,6 +48,8 @@
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.CodeLens" />
     <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.CSharp" />
+    <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.Implementation" />
+    <InternalsVisibleTo Include="Microsoft.VisualStudio.LanguageServices.LiveShare" />
```
Comment (jmarolf) [inline src/Workspaces/Remote/Core/Microsoft.CodeAnalysis.Remote.Workspaces.csproj:52]: why is this necessary?

Comment (sharwell) [reply src/Workspaces/Remote/Core/Microsoft.CodeAnalysis.Remote.Workspaces.csproj:52]: These are local projects in this solution. Added so they have access to the VisualStudioMefHostServices type in the shared layer. If we were willing to pick up a vs-mef dependency even lower (features layer) we could avoid this.

**Review by tmat — APPROVED** [review]
:shipit:

**Comment by sharwell** [comment]
@AArnott I would value your audit on this PR if you get the chance 😄

File: src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspaceManager.cs [hunk 2]
```diff
@@ -58,10 +60,50 @@ internal void InitializeAssetSource(IAssetSource assetSource)
         public IAssetSource? TryGetAssetSource()
             => _solutionAssetSource;
 
+        private static ComposableCatalog CreateCatalog(ImmutableArray<Assembly> assemblies)
+        {
+            var resolver = new Resolver(SimpleAssemblyLoader.Instance);
```
Comment (AArnott) [inline src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspaceManager.cs:65]: Why not use `Resolver.DefaultInstance`?

Comment (sharwell) [reply src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspaceManager.cs:65]: I used the one from other examples I had. If `DefaultInstance` works we can use it.
